### PR TITLE
MAINT: cleanup compat.py3k.py

### PR DESCRIPTION
**Pull request by hmaarrfk**
Supersedes https://github.com/numpy/numpy/pull/14093

Cleanups many if statements for python>=3.5

cc: @eric-wieser

### CODE REVIEW

**Comment by eric-wieser** [comment]
Probably worth updating the docstring at the top to indicate that:
* while numpy no longer supports python 2, downstream projects may still be using these shims
* the file is also needed for features that appeared in a later python 3

**Comment by charris** [comment]
> while numpy no longer supports python 2, downstream projects may still be using these shims

That would be my main concern. I haven't checked this PR, but none of the functions currently available for Python3 should disappear. We are also waiting for 1.19 development to start before doing Python2.7 cleanups, there are other pending PRs making a start on that. Note that internal cleanups are less problematical than changes to the public API.

**Comment by hmaarrfk** [comment]
> Note that internal cleanups are less problematical than changes to the public API.

This PR has been restricted to strictly internal cleanups, and the should not affect the public facing API at all. That said, it does fall in the realm of 2.7 cleanups, which are good to wait on.

File: numpy/compat/py3k.py [hunk 1]
```diff
@@ -1,9 +1,15 @@
...
-from __future__ import division, absolute_import, print_function
+While this file was originally intented for Python 2 -> 3 transition,
+it is now used to create a compatibility layer between different
+minor versions of Python 3.
 
+While the active version of numpy may not support a given version of python, we
+allow downstream libraries to continue to use these shims for forward
+compatibility with numpy while they transition their code to newer versions of
+Python.
+"""
```
Comment (eric-wieser) [inline numpy/compat/py3k.py]: Great comment, thanks - to the extent that I'd consider merging a PR with only this change in for 1.18.0, if it's worth your time to rebase.

**Comment by jarrodmillman** [comment]
@hmaarrfk It is finally time for 2.7 cleanups!  Would you mind rebasing on master?

**Comment by hmaarrfk** [comment]
done ;)

File: numpy/compat/py3k.py [hunk 2]
```diff
@@ -18,76 +18,49 @@
...
-
-if sys.version_info[0] >= 3:
-    import io
-
-    try:
-        import pickle5 as pickle
-    except ImportError:
-        import pickle
+from pathlib import Path, PurePath
+from importlib.machinery import SourceFileLoader
```
Comment (eric-wieser) [inline numpy/compat/py3k.py]: We should leave this import in `npy_load_module`, to save paying for an import of `importlib` at startup.

Comment (hmaarrfk) [reply numpy/compat/py3k.py]: should this be public??? I guess we are stuck with it. It seems to only be used in setup.py

File: numpy/compat/py3k.py [hunk 3]
```diff
@@ -134,69 +106,30 @@ def __exit__(self, *excinfo):
...
+    mod : module
 
-# backport abc.ABC
-import abc
-if sys.version_info[:2] >= (3, 4):
-    abc_ABC = abc.ABC
-else:
-    abc_ABC = abc.ABCMeta('ABC', (object,), {'__slots__': ()})
+    """
+    # Explitely lazy import this to avoid paying the cost
```
Comment (eric-wieser) [inline numpy/compat/py3k.py]:
```suggestion
    # Explicitly lazy import this to avoid paying the cost
```

File: numpy/compat/py3k.py (hunk 2 above)
Comment (eric-wieser) [reply numpy/compat/py3k.py]: We should probably replace its callers with direct usages of `importlib` within numpy, then deprecate it - but not as part of this patch.

**Comment by seberg** [comment]
Seems this has been finished for a while. Thanks, I will merge before we get conflicts with @sethtroisi work.

**Comment by hmaarrfk** [comment]
thanks!

**Pull request merged**